docs/usage/configuration-options.md

```diff
@@ -2200,22 +2200,64 @@ e.g.
 
 The `postUpgradeTasks` configuration consists of three fields:
 
-### commands
+### commands (postUpgradeTasks)
 
 A list of commands that are executed after Renovate has updated a dependency but before the commit is made.
 
-You can use variable templating in your commands if [`allowPostUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowpostupgradecommandtemplating) is enabled.
+You can use variable templating in your commands if [`allowUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowUpgradeCommandTemplating) is enabled.
 
-### fileFilters
+### fileFilters (postUpgradeTasks)
 
 A list of glob-style matchers that determine which files will be included in the final commit made by Renovate.
 
-### executionMode
+### executionMode (postUpgradeTasks)
 
 Defaults to `update`, but can also be set to `branch`.
 This sets the level the postUpgradeTask runs on, if set to `update` the postUpgradeTask will be executed for every dependency on the branch.
 If set to `branch` the postUpgradeTask is executed for the whole branch.
 
+## preUpgradeTasks
+
+<!-- prettier-ignore -->
+!!! note
+    Pre-upgrade tasks can only be used on self-hosted Renovate instances.
+
+Pre-upgrade tasks are commands that are executed by Renovate before a dependency has been updated but after the checkout on the dependency branch.
+The intention is to run any additional command line tools that would modify existing files or generate new files when a dependency changes.
+
+Each command must match at least one of the patterns defined in `allowedUpgradeCommands` (a global-only configuration option) in order to be executed.
+If the list of allowed tasks is empty then no tasks will be executed.
+
+e.g.
+
+```json
+{
+  "preUpgradeTasks": {
+    "commands": ["xcodegen generate"],
+    "fileFilters": [],
+    "executionMode": "update"
+  }
+}
+```
+
+The `preUpgradeTasks` configuration consists of three fields:
+
+### commands (preUpgradeTasks)
+
+A list of commands that are executed before Renovate has updated a dependency but after the checkout is made.
+
+You can use variable templating in your commands if [`allowUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowUpgradeCommandTemplating) is enabled.
+
+### fileFilters (preUpgradeTasks)
+
+A list of glob-style matchers that determine which files will be included in the final commit made by Renovate.
+
+### executionMode (preUpgradeTasks)
+
+Defaults to `update`, but can also be set to `branch`.
+This sets the level the preUpgradeTask runs on, if set to `update` the preUpgradeTask will be executed for every dependency on the branch.
+If set to `branch` the preUpgradeTask is executed for the whole branch.
+
```

Comment:
If `preUpgradeTasks` is for self-hosted only, then move the section into `docs/usage/self-hosted-configuration.md`. 😉

Reply:
If `postUpgradeTasks` is meant for self-hosted-only, then also move that section to `docs/usage/self-hosted-configuration.md`.

Reply:
Yes it is (https://github.com/renovatebot/renovate/pull/20423/files#diff-be0a39f342380493f17c380bbd694f1f05d18bdd8d7cecf174369084952102deL2177-L2181).